patterns/interpreter.md

```diff
@@ -0,0 +1,185 @@
+# Interpreter
+
+## Description
+
+If a problem occurs very often and requires long and repetitive steps to solve
+it, then the problem instances might be expressed in a simple language and an
+interpreter object could solve it by interpreting the sentences written in this
+simple language. Basically, for any kind of problems we define a [domain specific
+language](https://en.wikipedia.org/wiki/Domain-specific_language), then define
+a grammar for this language and design interpreter solving problem
+instances.
+
+## Motivation
+
+Our goal is translate simple mathematical expressions into
+postfix expressions (or [Reverse Polish notation](https://en.wikipedia.org/wiki/Reverse_Polish_notation))
+For simplicity, our expressions consists of ten digits `0`,...,`9` and two
+operations `+, -`. For example, the expression `2 + 4` is translated into `2 4  +`.
+
+## Context Free Grammars
+
+In order to understand the Interpreter pattern you need to know a little bit
+about formal language theory. If you already know what a grammar is feel free
+to skip this section, otherwise fasten your seat belts.
+
+Basicaly, a [Context Free Grammar (CFG)](https://en.wikipedia.org/wiki/Context-free_grammar)
+defines a set of rules that describe all possible strings over some finite
+set of symbols, formally called alphabet. These strings usually have some well
+defined structure. For example, strings of all balanced parentheses:
+
+```ignore
+(), (()), ()(), ()(()), ..
+```
+
+Formally a CFG consists of terminal symbols, non-terminal symbols, a start
+(non-terminal) symbol, and production rules.
+
+In the previous example we have two terminal symbols `(` and `)`. We define a
+single non-terminal (at the same time the start) symbol `S` and production rules
+as:
+
+```ignore
+S -> SS
+S -> (S)
+S -> ()
+```
+
+So, in order to generate a string with balanced parentheses we repeatedly apply
+production rules. For example, to derive `()()` we apply the following rules:
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
+## Context Free Grammar for our problem
+
+Recall that our task is translate infix expressions into postfix ones. Let's
+define a context free grammar for a set of infix expressions over `0,...,9, +,-`,
+where:
+
+- terminal symbols: `0,...,9, +,-`
+- non-terminal symbols: `exp, term`
+- start symbol is `exp`
+- and the following are production rules
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed depending on what we are going
```

Comment:
I don't understand how "in fact" is connected with the rest.

Reply:
@MarcoIeni This CFG is not suitable yet to parse with a recursive descent parser. It has left recursion which should be eliminated. But I didn't go into details.

Reply:
Ok, let's see if other reviewers find that there is a problem here.
If nobody replies within a week or so we can mark this as resolved.

Reply:
@MarcoIeni We could remove this part if it is confusing. I put it here just in case someone asks "how did they come up with this Rust code from the gramma?". In fact implementation is up to implementer. I'd personally prefer manually coding if the grammar is "simple" enough to be addressed with a recursive descent parser. Another one could prefer tools such as Yacc, Bison.... or just use table. You know this is outside of our scope. All of them could be found in a book on compilers or specifically parsers. As I mentioned, this pattern is NOT at all about parsers. So, if you feel this "in fact" part does not fit here, let's remove. Or you could suggest another wording.

Reply:
I wrapped the text block under it into a "Note", so it's more seen as some additional explanation. Should be fine imho.